core/rawdb/freezer_meta.go

```diff
@@ -0,0 +1,236 @@
+	deleted uint64 // The number of items that have been removed from the table
+	hidden  uint64 // The number of items that have been hidden in the table
+}
+
+// newMetadata initializes the metadata object with the given parameters.
+func newMetadata(tailId uint32, deleted uint64, hidden uint64) *freezerTableMeta {
+	return &freezerTableMeta{
+		version: freezerVersion,
+		tailId:  tailId,
+		deleted: deleted,
+		hidden:  hidden,
+	}
+}
+
+// encodeMetadata encodes the given parameters as the freezer table metadata.
+func encodeMetadata(meta *freezerTableMeta) ([]byte, error) {
+	buffer := new(bytes.Buffer)
+	if err := rlp.Encode(buffer, meta.version); err != nil {
+		return nil, err
+	}
+	if err := rlp.Encode(buffer, meta.tailId); err != nil {
+		return nil, err
+	}
+	if err := rlp.Encode(buffer, meta.deleted); err != nil {
+		return nil, err
+	}
+	if err := rlp.Encode(buffer, meta.hidden); err != nil {
+		return nil, err
+	}
+	buffer.Write(make([]byte, metaLength-buffer.Len())) // Right pad zero bytes to the specified length
+	return buffer.Bytes(), nil
+}
+
+// decodeMetadata decodes the freezer-table metadata from the given
+// rlp stream.
+func decodeMetadata(r *rlp.Stream) (*freezerTableMeta, error) {
+	var version uint16
+	if err := r.Decode(&version); err != nil {
+		return nil, err
+	}
+	if version != freezerVersion {
+		return nil, newIncompatibleError(version)
+	}
+	var tailId uint32
+	if err := r.Decode(&tailId); err != nil {
+		return nil, err
+	}
+	var deleted, hidden uint64
+	if err := r.Decode(&deleted); err != nil {
+		return nil, err
+	}
+	if err := r.Decode(&hidden); err != nil {
+		return nil, err
+	}
+	return newMetadata(tailId, deleted, hidden), nil
+}
+
+// storeMetadata stores the metadata of the freezer table into the
+// given index file.
+func storeMetadata(index *os.File, meta *freezerTableMeta) error {
+	encoded, err := encodeMetadata(meta)
+	if err != nil {
+		return err
+	}
+	if _, err := index.WriteAt(encoded, 0); err != nil {
+		return err
+	}
+	return nil
+}
+
+// loadMetadata loads the metadata of the freezer table from the
+// given index file. Return the error if the version of loaded
+// metadata is not expected.
+func loadMetadata(index *os.File) (*freezerTableMeta, error) {
+	stat, err := index.Stat()
+	if err != nil {
+		return nil, err
+	}
+	if stat.Size() < metaLength {
+		return nil, newIncompatibleError(0)
+	}
+	buffer := make([]byte, metaLength)
+	if _, err := index.ReadAt(buffer, 0); err != nil {
+		return nil, err
+	}
+	return decodeMetadata(rlp.NewStream(bytes.NewReader(buffer), 0))
+}
+
+// upgradeV0TableIndex extracts the indexes from version-0 index file and
+// encodes/stores them into the latest version index file.
+func upgradeV0TableIndex(index *os.File) error {
+	// Create a temporary offset buffer to read indexEntry info
+	buffer := make([]byte, indexEntrySize)
+
+	// Read index zero, determine what file is the earliest
+	// and how many entries are deleted from the freezer table.
```

Comment:
Is this needed? Before v1, we haven't had tail deletion implemented

Reply:
Right, will fix it